RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco2D.cc

```diff
@@ -0,0 +1,654 @@
+      xe = 0.f;
+      ysum[i] = 0.f;
+      for(j=1; j<nclusx+1; ++j) {
+         float xpitch = xsize;
+         if(xdouble[j-1]) {xpitch += xsize;}
+         // Truncate pixel charges
+         if(cluster(j-1,i-1) > maxpix) {
+            clusxy[j][i] = maxpix;
+         } else {
+            clusxy[j][i] = cluster(j-1,i-1);
+         }
+         if(clusxy[j][i] > 0.) {
+            ysum[i] += clusxy[j][i];
+            //          xq += (xe + xpitch/2.)*clusxy[j][i];
+            if(j < jmin) {jmin = j; xlow0 = xe;}
+            if(j > jmax) {jmax = j; xhigh0 = xe+xpitch;}
+            if(i < imin) {imin = i; ylow0 = ye;}
+            if(i > imax) {imax = i; yhigh0 = ye+ypitch;}
+            indexxy[0][npixel] = j;
+            indexxy[1][npixel] = i;
+            pixel[npixel] = clusxy[j][i];
+            ++npixel;
+         }
+         xe += xpitch;
+      }
+      //      qtot += ysum[i];
+      ye += ypitch;
+   }
+   ysum[nclusy+1] = 0.f;
+   ypos[nclusy+1] = yhigh0+0.5f*ysize;
+   
+   //   float xbaryc = xq/qtot;
+   
+   // Quit if only one pixel in cluster
+   
+      if(npixel < 2 ) {
+//         LOGDEBUG("SiPixelTemplateReco2D") << "2D fit not possible with single pixel" << ENDL;
+         return 1;
+      }
+   
+   
+   //   if(jmax-jmin == 0 ) {
+   //      LOGDEBUG("SiPixelTemplateReco2D") << "2D fit not possible with single x-pixel" << ENDL;
+   //      return 1;
+   //   }
+   
+   // Quit if only one pixel in cluster
+   
+   //   if(imax-imin == 0) {
+   //      LOGDEBUG("SiPixelTemplateReco2D") << "2D fit not possible with single y-pixel" << ENDL;
+   //      return 1;
+   //   }
+   
+   // Next, calculate the error^2 [need to know which is the middle y pixel of the cluster]
+   
+   int ypixoff = T2HYP1 - (imin+imax)/2;
+   for(k=0; k<npixel; ++k) {
+      i = indexxy[1][k];
+      int ypixeff = ypixoff + i;
+      templ2D.xysigma2(pixel[k], ypixeff, sigma2[k]);
+   }
+   
+   // Next, find the half-height edges of the y-projection and identify any
+   // missing columns to remove from the fit
+   
+   int imisslow = -1, imisshigh = -1, jmisslow = -1, jmisshigh = -1;
+   float ylow = -1.f, yhigh = -1.f;
+   float hmaxpix = fracpix*templ2D.sxymax();
+   for(i=imin; i<=imax; ++i) {
+      if(ysum[i] > hmaxpix && ysum[i-1] < hmaxpix && ylow < 0.f) {
+         ylow = ypos[i-1] + (ypos[i]-ypos[i-1])*(hmaxpix-ysum[i-1])/(ysum[i]-ysum[i-1]);
+      }
+      if(ysum[i] < q50) {
+         if(imisslow < 0) imisslow = i;
+         imisshigh = i;
+      }
+      if(ysum[i] > hmaxpix && ysum[i+1] < hmaxpix) {
+         yhigh = ypos[i] + (ypos[i+1]-ypos[i])*(ysum[i]-hmaxpix)/(ysum[i]-ysum[i+1]);
+      }
+   }
+   if(ylow < 0.f || yhigh < 0.f) {
+      ylow = ylow0;
+      yhigh = yhigh0;
+   }
+   
+   float templeny = templ2D.clsleny();
+   deltay = templeny - (yhigh - ylow)/ysize;
+   
+   // Calculate the corrected distance from the ylow and yhigh to the cluster center
+   
+//   float delycor = templ2D.delyavg();

```

Comment:
Are these commented out lines needed? If so, please add another line of comment which explain why; otherwise just remove them

Reply:
I don't know.  I'll need to discuss with Morris.  I propose to postpone this until later.